pkg/controller/pod/pod_controller.go

```diff
@@ -250,3 +264,106 @@ func (r *ReconcilePod) createK8sObject(obj *unstructured.Unstructured) error {
 	}
 	return nil
 }
+
+func (r *ReconcilePod) recreateNodeAgentPodsIfInvalidResolvConf(
+	resName string) error {
+	podsInCLB, err := identifyPodsInCLBDueToInvalidResolvConf(r.client)
+	if err != nil {
+		log.Error(err, "Could not identify if any pod is in CLB because "+
+			"of invalid resolv.conf")
+		return err
+	}
+	if len(podsInCLB) > 0 && !deletePods(podsInCLB, r.client) {
+		err := errors.New("Error occured while trying to restart pods in " +
+			"CLB because of invalid resolv.conf")
+		log.Error(err, "")
+		return err
+	}
+	return nil
+}
+
+func identifyPodsInCLBDueToInvalidResolvConf(c client.Client) (
+	[]corev1.Pod, error) {
+	var podsInCLB []corev1.Pod
+	podList := &corev1.PodList{}
+	nodeAgentLabelSelector := labels.SelectorFromSet(
+		map[string]string{"component": operatortypes.NsxNodeAgentDsName})
+	err := c.List(context.TODO(), podList, &client.ListOptions{
+		LabelSelector: nodeAgentLabelSelector})
+	if err != nil {
+		return nil, err
+	}
+	for _, pod := range podList.Items {
+		if isNodeAgentContainerInCLB(&pod) {
+			nodeAgentLogs, err := getContainerLogsInPod(
+				&pod, operatortypes.NsxNodeAgentContainerName)
+			if err != nil {
+				log.Error(err, "Error occured while getting container logs")
+				return nil, err
+			}
+			if strings.Contains(
+				nodeAgentLogs, "Failed to connect k8s apiserver") {
+				log.Info(fmt.Sprintf(
+					"Pod %v is in CLB because of invalid resolv.conf. "+
+						"It shall be restarted", pod.Name))
+				podsInCLB = append(podsInCLB, pod)
+			}
+		}
+	}
+	return podsInCLB, nil
+}
+
+func isNodeAgentContainerInCLB(pod *corev1.Pod) bool {
+	for _, containerStatus := range pod.Status.ContainerStatuses {
+		if containerStatus.Name == operatortypes.NsxNodeAgentContainerName {
+			if containerStatus.State.Waiting != nil &&
+				containerStatus.State.Waiting.Reason == "CrashLoopBackOff" {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+var getContainerLogsInPod = func(pod *corev1.Pod, containerName string) (
+	string, error) {
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		return "", err
+	}
+	clientSet, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return "", err
+	}
+	podLogOptions := &corev1.PodLogOptions{
+		Container: operatortypes.NsxNodeAgentContainerName,
+		Previous:  true}
```

Comment:
do you think it might be beneficial to set sinceSeconds to a reasonable value to avoid fetching too much data?

Reply:
these are crashed container logs. the error happens within 4-5 seconds after a container is started. Not 100% confident if setting it to (current time - 30 seconds) would guarantee correct behaviour because of the exponential backoff. What do you think?

Reply:
fetching the last 50 log lines.